kotlin-analysis/src/main/kotlin/org/jetbrains/dokka/analysis/KotlinAnalysis.kt

```diff
@@ -34,20 +62,53 @@ class DokkaAnalysisConfiguration(
 )
 fun KotlinAnalysis(context: DokkaContext): KotlinAnalysis = KotlinAnalysis(context.configuration.sourceSets, context.logger)
 
-interface KotlinAnalysis : SourceSetDependent<EnvironmentAndFacade> {
-    override fun get(key: DokkaSourceSet): EnvironmentAndFacade
-    operator fun get(sourceSetID: DokkaSourceSetID): EnvironmentAndFacade
+/**
+ * First child delegation. It does not close [parent].
+ */
+abstract class KotlinAnalysis(
+    val parent: KotlinAnalysis? = null
+) : Closeable {
+
+    operator fun get(key: DokkaSourceSet): AnalysisContext {
+        return find(key)
+            ?: parent?.get(key)
+            ?: throw IllegalStateException("Missing EnvironmentAndFacade for sourceSet ${key}")
+    }
+
+    operator fun get(sourceSetID: DokkaSourceSetID): AnalysisContext {
+        return find(sourceSetID)
+            ?: parent?.get(sourceSetID)
+            ?: throw IllegalStateException("Missing EnvironmentAndFacade for sourceSet ${sourceSetID.sourceSetName}")
+    }
+
+    protected abstract fun find(sourceSetID: DokkaSourceSetID): AnalysisContext?
+    protected abstract fun find(sourceSet: DokkaSourceSet): AnalysisContext?
```

Comment:
`DokkaSourceSetID` can not be unique.  Anyway, we do not check it. And it is a more efficient search.

Reply:
Why do you mean by `more efficient search`? It's an abstract method :) If you're talking about the implementation - maybe, and maybe not. `DokkaSourceSetImpl` has 20+ fields, half of which are collections. I'm not sure comparing it by `equals()` is much faster than iterating over a list once and comparing strings.

Also, can you please give an example of when it's not unique? How do you then identify a unique `DokkaSourceSet`? I'm asking because we rely on `DokkaSourceSetID` for source set dependencies, which might lead to bugs in this case

Reply:
> It's an abstract method :)

Yes, you could have a more efficient implementation than `find(sourceSet.sourceSetID)`.

> I'm not sure comparing it by equals() is much faster than iterating over a list once and comparing strings.

`equals()` can be not needed since we have a hash table. Otherwise, they will be compared by a reference first (our cases).

> can you please give an example of when it's not unique?

https://github.com/Kotlin/dokka/pull/2755/commits/833a9dc41a78a97be78c96cb88f9df82f2e36eb3

Reply:
It can surely be non unique in our test cases where we set the values manually. Can it be non unique in a real production library/project? If it can, I don't think this method should exist at all

> Yes, you could have a more efficient implementation than find(sourceSet.sourceSetID)

I'm sorry, I'm not buying it. What stops the implementation from having a `Map<DokkaSourceSetID, value>`? How is it "more efficient" on the interface level? Even if so, it doesn't make sense to have two methods that do the same thing from the outside, but one is more efficient than the other - how does the user know which one to call? It's a leaky abstraction

Reply:
> How is it "more efficient" on the interface level? 

I left it due to correctness rather than performance. 

I think it is better to remove 
``protected abstract fun find(sourceSetID: DokkaSourceSetID): AnalysisContext?``
Objections?

Reply:
Sure 👍

Reply:
Sorry, you've deprecated it with the reason `sourceSetID can be non unique`, which is a serious statement, but I still don't understand in which situations in can be non unique. Can you please give an example?

I had a closer look at the [commit that you linked above](https://github.com/Kotlin/dokka/commit/833a9dc41a78a97be78c96cb88f9df82f2e36eb3), and it's definitely a configuration error that has not been fixed. In the [test source set builder](https://github.com/Kotlin/dokka/blob/8c218ff4dd5f970233c43845c19299fc74256389/core/test-api/src/main/kotlin/testApi/testRunner/TestDokkaConfigurationBuilder.kt#L112), the ID is constructed as `DokkaSourceSetID(moduleName, name)`, which by default is `(root, main)`. In the commit, `displayName` was added, which helps in `DokkaSourceSet#equals`, but there are still two source sets with the same name `main`, which is invalid configuration.